Add basic support for explicit copy operations

psalz opened the pull request:
This adds support for the most important explicit copy operations, namely transferring data between all combinations of host and device accessors for global buffers, as well as raw and shared host pointers. This means that e.g. images are not yet supported.

1D and 2D copying is supported on all platforms, while 3D copying is currently only supported on CUDA. This is because HIP currently [does not provide](https://github.com/ROCm-Developer-Tools/HIP/blob/master/docs/markdown/CUDA_Runtime_API_functions_supported_by_HIP.md) a `hipMemcpy3DAsync`. It does have a `hipMemcpy3D` which we could fall back to, however that uses a slightly different parameter struct compared to its CUDA counterpart, and as I don't have access to a working HCC setup, I couldn't test this - so I omitted it for now.

In order to support copying of subranges at arbitrary locations within source and target buffers, I also added proper handling of ranges and offsets for `accessor`. Unfortunately I didn't have time to add explicit test cases for the accessor API, however it should at least partially be covered by the new and existing test cases.

On that note, going through the spec I couldn't really determine what the exact semantics for copying between two accessors that both have an offset and a range should be. I assume (and that is what I implemented) that the first byte copied from `src` is at `src.pointer + linear_id(src.offset, src.buffer_range)`, and is written to `dest.pointer + linear_id(dest.offset, dest.buffer_range)`. Also the total number of copied elements is determined by `src.range`, which is why `dest.range` must at least be sized equally. The reason I'm confused about this is that ComputeCpp seems to be doing something different (*what* exactly I couldn't really figure out, I certainly doesn't seem 100% right -- which however could be due to the fact that I'm running their experimental PTX backend...). Since triSYCL doesn't support `copy()` yet, I couldn't verify the behavior with a third party.

In addition to that, the template parameterization given in the spec seems to have some issues, as e.g. this

```c++
template <typename T, int dim, access::mode mode, access::target tgt>
void copy(accessor<T, dim, mode, tgt> src, accessor<T, dim, mode, tgt> dest)
```
would enforce that both `src` and `dest` have the same access mode and target, which doesn't make much sense.

---

On a more technical level, I'm using `buffer_impl::register_external_access` whenever a host accessor is involved to ensure that `copy()` operations are correctly considered when determining buffer dependencies (device accessors are already covered by the existing mechanism involving `handler::_detail_add_access`). From looking at the debug output of the task graph I think this works, but please have a look to see if you can spot any issues with that.

The tests mostly focus on the overload of `copy()` that uses two accessors, as it is the most involved one. Ultimately however I tried to share as much code as possible between all variants, and especially the raw and shared pointer variants use the exact same code.

A known limitation for now is that since hipSYCL currently only tracks buffer versions for entire buffers, performing an explicit copy using a discard_* access mode into a subrange of a buffer will invalidate the entire buffer range (but that issue is not exclusive to explicit copy operations).

illuhad commented on include/CL/sycl/handler.hpp:
```diff
@@ -597,18 +659,221 @@ void set_args(Ts &&... args);
+    // It looks like HIP doesn't provide a hipMemcpy3DAsync as of April 2019.
+    // See https://github.com/ROCm-Developer-Tools/HIP/blob/master/docs/markdown/CUDA_Runtime_API_functions_supported_by_HIP.md
+    // TODO: We could fall back to hipMemcpy3D on HCC (however, that requires a different parameter struct).
+    throw feature_not_supported{"3D copy() is currently not supported on "
```
okay!

illuhad commented:
> This is because HIP currently does not provide a hipMemcpy3DAsync. It does have a hipMemcpy3D which we could fall back to, however that uses a slightly different parameter struct compared to its CUDA  counterpart, and as I don't have access to a working HCC setup, I couldn't test this - so I omitted it for now.

`hipMemcpy3D` should also be available in HIP's CUDA backend, so the functionality of such a solution could already be tested with a CUDA stack ;)

 >I implemented) that the first byte copied from src is at src.pointer + linear_id(src.offset, src.buffer_range), and is written to dest.pointer + linear_id(dest.offset, dest.buffer_range). Also the total number of copied elements is determined by src.range, which is why dest.range must at least be sized equally. The reason I'm confused about this is that ComputeCpp seems to be doing something different (what exactly I couldn't really figure out, I certainly doesn't seem 100% right -- which however could be due to the fact that I'm running their experimental PTX backend...). Since triSYCL doesn't support copy() yet, I couldn't verify the behavior with a third party.

This is also the behavior that I'd expect. In fact, I would say that the spec requires this. For the accessor to accessor case, it says:
> [...] Copies the contents of the memory object accessed by src into the memory object accessed by dest. [...]

And of course, the memory regions accessed by the buffers are given by what you implemented. To be sure, could you perhaps try with ComputeCpp on Intel CPU/GPU to check if it's really just the PTX backend?

> In addition to that, the template parameterization given in the spec seems to have some issues, as e.g. this [...] would enforce that both src and dest have the same access mode and target, which doesn't make much sense.

I think it makes kind of sense for the target. If you want to copy from host->device you can just create a command group with an accessor that requests the memory. SYCL runtime will then make sure that the memory is copied to the device, if necessary. For the other direction you can use `handler::update_host()`. So, the common use case will be to copy between same targets (And e.g. copying between an image and a buffer probably doesn't make much sense either).
For the access mode I'm inclined to say that it is actually wrong to use the same modes. If you have two `read` accessors, and copy between them, all hell may break lose because the synchronization rules are more relaxed for read-only accesses than for write accesses. And the SYCL runtime does not treat the `dest` accessor as a write in this case.
One solution would be to automatically upgrade accessors to read-write accessors if they are used in a copy. However, I find that solution counterintuitive (if the user has explicitly provided a read-only accessor) and I'm not really sure that's how it was intended. I would suggest that we require sensible, different modes for the accessors in hipSYCL to ensure that dependencies are calculated correctly.
I can also ask inside the Khronos SYCL committee to clarify this.

> On a more technical level, I'm using buffer_impl::register_external_access whenever a host accessor is involved to ensure that copy() operations are correctly considered when determining buffer dependencies (device accessors are already covered by the existing mechanism involving handler::_detail_add_access). From looking at the debug output of the task graph I think this works, but please have a look to see if you can spot any issues with that.

`register_external_access` only adds the dependency, but it does not modify buffer versioning. If we require that dest buffers should be of a write access mode, as discussed above, I think this *should* be fine since the construction of the buffer will already take care of that.

> A known limitation for now is that since hipSYCL currently only tracks buffer versions for entire buffers, performing an explicit copy using a discard_* access mode into a subrange of a buffer will invalidate the entire buffer range (but that issue is not exclusive to explicit copy operations).

Yeah, subranges are not yet properly supported anyway - to do that properly we need a data structure to efficiently track buffer versions at a sub-buffer level, in arbitrary granularity. Future work :)

illuhad commented:
> In addition to that, the template parameterization given in the spec seems to have some issues, as e.g. this [...] would enforce that both src and dest have the same access mode and target, which doesn't make much sense.

New information: Turns out that's a known typo in the spec. You can use the following signature:
```cpp
 template <typename T_src, int dim_src, access::mode mode_src, access::target tgt_src, access::placeholder isPlaceholder_src,
            typename T_dest, int dim_dest, access::mode mode_dest, access::target tgt_dest, access::placeholder isPlaceholder_dest>
  void copy(accessor<T_src, dim_src, mode_src, tgt_src, isPlaceholder_src> src,
            accessor<T_dest, dim_dest, mode_dest, tgt_dest, isPlaceholder_dest> dest);
```
However, I propose that in order to not have the full complexity at once, we only implement the cases where the target is the same for now (generalizing this can be another PR). We can also ignore different type or placeholder properties for now and see what ends up in SYCL-next spec.

Additionally:
* local is not allowed as target
* source accessor can be read or read_write
* destination accessor can be write, read_write, discard_write, discard_read_write

psalz commented:
Thanks for the feedback and for clarifying with the committee!

I've moved some of the helper functions around, as per your suggestions.

> Have you encountered any issues with the CPU backend? I think many of those hipMemcpy functions are not yet implemented in hipCPU, so we should make sure to implement them ASAP.

Good point, frankly I just assumed it would be implemented as the declarations do exist in hipCPU. But yes, I get linker errors for the missing functions. Since they're only being used inside template functions however, I think this at least shouldn't break any existing hipSYCL code using the CPU backend.

> `hipMemcpy3D` should also be available in HIP's CUDA backend, so the functionality of such a solution could already be tested with a CUDA stack ;)

Unfortunately no -- while CUDA uses the same [argument struct](http://developer.download.nvidia.com/compute/cuda/2_3/toolkit/docs/online/structcudaMemcpy3DParms.html) for both `cudaMemcpy3D` and `cudaMemcpy3DAsync`, the [HCC variant](https://github.com/ROCm-Developer-Tools/HIP/blob/78fd844a169e9c3a51c3da7ddd1987f5fd9df369/include/hip/hcc_detail/driver_types.h#L245) looks completely different.

> To be sure, could you perhaps try with ComputeCpp on Intel CPU/GPU to check if it's really just the PTX backend?

So I did some more testing on my Windows machine with an RX 580 and i7-2600k. Both produce a faulty result when copying part of a larger 2D buffer into a smaller buffer with an offset. As far as I can tell it does work correctly for their host device, which leads me to believe that they have some error in the way they call `clEnqueueCopyBufferRect` (if that is what they're using).

> However, I propose that in order to not have the full complexity at once, we only implement the cases where the target is the same for now (generalizing this can be another PR). We can also ignore different type or placeholder properties for now and see what ends up in SYCL-next spec.

Well, given that I've already implemented part of this, are you suggesting I should take it back out? That would be unfortunate, as I'm currently relying on `copy()` being able to copy between different targets in my own project.

psalz commented on include/CL/sycl/handler.hpp:
```diff
@@ -113,6 +115,47 @@ __global__ void parallel_for_workgroup(Function f,
+T* get_raw_pointer(T* ptr) { return ptr; }
+
+template <typename WrappedPtr>
+auto get_raw_pointer(const WrappedPtr& ptr) { return ptr.get(); }
```
Done!

psalz commented on include/CL/sycl/handler.hpp:
```diff
@@ -113,6 +115,47 @@ __global__ void parallel_for_workgroup(Function f,
+auto get_raw_pointer(const WrappedPtr& ptr) { return ptr.get(); }
+
+template <typename T, int dim, access::mode mode, access::target tgt>
+size_t get_pointer_offset(const sycl::accessor<T, dim, mode, tgt>& acc)
```
Done!

psalz commented on include/CL/sycl/handler.hpp:
```diff
@@ -113,6 +115,47 @@ __global__ void parallel_for_workgroup(Function f,
+}
+
+template <int dimsOut, int dimsIn>
+sycl::range<dimsOut> range_cast(const sycl::range<dimsIn>& other)
```
Done!

psalz commented on include/CL/sycl/handler.hpp:
```diff
@@ -113,6 +115,47 @@ __global__ void parallel_for_workgroup(Function f,
+    if(dstTgt == access::target::global_buffer) return hipMemcpyHostToDevice;
+    if(dstTgt == access::target::host_buffer) return hipMemcpyHostToHost;
+  }
+  assert(false && "Unsupported / unimplemented access targets");
```
Throwing an exception would require making this function non-`constexpr`. That in itself wouldn't be a big deal, but generally I'd say this assertion is more of a sanity check, as any invalid targets provided by the user should at this point already have been cought by the `handler:::validate_copy_[src|dest]_accessor` functions!

illuhad replied:
Okay, I see. In that case `assert` seems like a good choice.

psalz commented:
Oops, forgot to commit the new `detail/util.hpp` file - fixed.

illuhad commented:
Thank you!

> Good point, frankly I just assumed it would be implemented as the declarations do exist in hipCPU. But yes, I get linker errors for the missing functions. Since they're only being used inside template functions however, I think this at least shouldn't break any existing hipSYCL code using the CPU backend.

Okay. I'll create an issue to add these functions to make sure that it's not forgotten.

> Unfortunately no -- while CUDA uses the same argument struct for both cudaMemcpy3D and cudaMemcpy3DAsync, the HCC variant looks completely different.

hm, this seems like a bug in HIP because what good is a portability layer if it doesn't accept uniform parameters across platforms. Perhaps we should file a pull request to implement some uniform behaviour in HIP?

> So I did some more testing on my Windows machine with an RX 580 and i7-2600k. Both produce a faulty result when copying part of a larger 2D buffer into a smaller buffer with an offset. As far as I can tell it does work correctly for their host device, which leads me to believe that they have some error in the way they call clEnqueueCopyBufferRect (if that is what they're using).

Great, so let's assume that it's a ComputeCpp problem and not our problem :)

> Well, given that I've already implemented part of this, are you suggesting I should take it back out? That would be unfortunate, as I'm currently relying on copy() being able to copy between different targets in my own project.

Of course you don't have to remove stuff that the spec requires and that already works :) I hadn't realized that this is already implemented. Sorry!
